File: LeetCode/Study Plan/Practice Problems/Hard/Basic Calculator/Cpp/basic_calculator.cc
#include <iostream>
#include <stack>
#include <cmath>

typedef std::stack<std::string> ss;

int eval(ss &st) {
    bool f = false;
    std::string h = st.top();
    if (h=="+" || h=="-") f = true;
    if (st.empty() || f) {
        st.push("0");
    }
    int res = std::stoi(st.top());
    st.pop();
    while (!st.empty() && st.top()!=")") {
        std::string sign = st.top();
        st.pop();
        if (sign == "+") res += std::stoi(st.top());
        else res -= std::stoi(st.top());
        st.pop();
    }
    return res;
}

int basic_calculator(std::string s) {
    ss st;
    int op = 0;
    int n = 0;
    for (int i=s.size()-1; i>=0; i--) {
        if (std::isdigit(s[i])){
            op = pow(10, n)*(s[i]-'0') + op;
            n++;
        }else if (s[i] != ' '){
            if (n != 0) {
                st.push(std::to_string(op));
                n = 0;
                op = 0;
            }
            if (s[i] == '(') {
                int res = eval(st);
                st.pop();
                st.push(std::to_string(res));
            }else {
                std::string t = "";
                t += s[i];
                st.push(t);
            }
        }
    }
    if (n != 0) {
        st.push(std::to_string(op));
    }
    return eval(st);
}

int main() {
    std::cout<<basic_calculator("(5-(1+(5)))")<<"\n";
}

File: LeetCode/Hard/Basic Calculator/Go/basic_calculator.go
package main

import (
	"fmt"
	"math"
	"strconv"
)

func basic_calculator(s string) int {
	stack := []string{}
	op := 0
	n := 0
	for i := len(s) - 1; i >= 0; i-- {
		ch := s[i]
		if ch >= '0' && ch <= '9' {
			op = int(math.Pow(10, float64(n)))*int(ch-'0') + op
			n++
		} else if ch != ' ' {
			if n != 0 {
				stack = append(stack, strconv.Itoa(op))
				op = 0
				n = 0
			}
			if ch == '(' {
				res := eval(&stack)
				stack = stack[:len(stack)-1]
				stack = append(stack, strconv.Itoa(res))
			} else {
				stack = append(stack, string(s[i]))
			}
		}
	}
	if n != 0 {
		stack = append(stack, strconv.Itoa(op))
	}
	return eval(&stack)
}

func eval(stack *[]string) int {
	n := len(*stack)
	f := false
	_, err := strconv.Atoi((*stack)[n-1])
	if err != nil {
		f = true
	}
	if n == 0 || f {
		*stack = append(*stack, "0")
	}
	res, _ := strconv.Atoi((*stack)[len(*stack)-1])
	*stack = (*stack)[:len(*stack)-1]
	for len(*stack) > 0 && (*stack)[len(*stack)-1] != ")" {
		sign := (*stack)[len(*stack)-1]
		*stack = (*stack)[:len(*stack)-1]
		if sign == "+" {
			v, _ := strconv.Atoi((*stack)[len(*stack)-1])
			res += v
		} else {
			v, _ := strconv.Atoi((*stack)[len(*stack)-1])
			res -= v
		}
		*stack = (*stack)[:len(*stack)-1]
	}
	return res
}

func main() {
	fmt.Println(basic_calculator("1 + 1"))
}
